### PR TITLE
[Docs]: update reduce_max/min/sum docstring, add dtype coverage tests and API docs

### PR DESCRIPTION
**Pull request by 2Sheldonlaw**
## 改动内容

针对 `T.reduce_max` / `T.reduce_min` / `T.reduce_sum` 三个 API 完成文档校验、测试用例补充和 docstring 更新。

### 1. Docstring 更新（修改）

- `tilelang/language/reduce_ascend.py` 中三个 reduce 函数的 docstring
- 在原版基础上补充，不删除原有内容：
  - dtype 支持（float16, float32）
  - 内存位置（UB, `T.alloc_ub`）
  - clear 参数语义（True: 初始化 out; False: 累加/比较）
  - reduce_sum 补充 float16 溢出说明

### 2. 测试用例（新增）

- `testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py`
- **32 个用例**（31 PASSED + 1 SKIPPED），不重复 `elementwise.py` / `narrow_reduce.py` 中已有的测试
- LP 分层标记：PR 合入时跑 6 个，每日 CI 跑 32 个

| 测试函数 | 用例数 | PR执行 | LP | 覆盖内容 |
|---------|:---:|:---:|:---:|---------|
| `test_reduce_clear_false_2d` | 12 | 2 | 10 | 2D clear=False 累加模式（仓库仅有 1D 的 clear=False 测试） |
| `test_reduce_unsupported_dtype` | 6 | 2 | 4 | bfloat16 应编译失败（仓库无异常边界测试） |
| `test_reduce_out_shape_mismatch` | 2 | 1 | 1 | out shape 不匹配应 ValueError |
| `test_reduce_dim_not_int` | 2 | 0 | 2 | dim 为 tuple 应 TypeError |
| `test_reduce_real_shape_wrong_length` | 2 | 0 | 2 | real_shape 长度≠2 应 ValueError |
| `test_reduce_real_shape_exceed_extent` | 2 | 0 | 2 | real_shape 超 buffer 应 ValueError |
| `test_reduce_1d_invalid_dim` | 2 | 0 | 2 | 1D dim=1 应 ValueError |
| `test_reduce_2d_invalid_dim` | 2 | 1 | 1 | 2D dim=2 应 ValueError |
| `test_reduce_rank_too_high` | 2 | 0 | 2 | 4D buffer 应 ValueError |

LP 标记策略（参照 mentor 的 tail_block PR）：
- reduce kind：sum 默认执行，max/min 标 `low_priority`（三者走相同校验路径）
- dtype：float32 默认执行，float16 标 `low_priority`
- target：ascendc/pto 均默认执行
- 异常边界测试：保留 2 个代表性用例默认执行（`out_shape_mismatch` / `2d_invalid_dim` 的 ascendc 版本，验证 validator 报错能力），其余标 `low_priority`（异常校验发生在 Python validator 层，与后端无关）

### 3. API 文档（新增）

- `docs/api_docs/T.reduce_max.md`
- `docs/api_docs/T.reduce_min.md`
- `docs/api_docs/T.reduce_sum.md`

校验完成版文档，包含：功能说明 / 函数原型 / 参数说明 / DataType 支持 / Shape 支持 / dim 说明 / 约束条件 / 示例代码。

## 测试结果

- **31 PASSED + 1 SKIPPED**（CI 全绿，0 FAILED）

### 1 个 SKIPPED

| 用例 | 原因 |
|------|------|
| `test_reduce_clear_false_2d[ascendc-float16-sum]` | CANN `ReduceSum<half, AR>` 不支持 half（clear=False 路径无 `reduce_sum_half` workaround），pto 后端正常通过 |

## 已知限制（文档约束已写明）

| 限制 | 文档位置 | 说明 |
|------|---------|------|
| 3D 不支持 | 三个文档 2.3.2 Shape | 精度异常，暂不支持 |
| BufferRegion pto 不支持 | 三个文档约束 8 | pto codegen `GetSliceInfo` 对切片行数计算有误 |
| float16 列归约 ascendc 不支持 | `T.reduce_sum.md` 约束 9 | `reduce_sum_half` 的 `srcRepStride` 无法表达 2 字节列步进 |
| float16 clear=False ascendc 不支持 | `T.reduce_sum.md` 约束 10 | CANN `ReduceSum<half>` 被 `static_assert` 拒绝 |

## 校验过程中发现的 framework bug

本次校验过程中发现 5 个 framework bug，均不影响本 PR 合入（已通过 skip/删除测试/文档约束处理），详细根因分析记录在 `api/bug_investigation/reduce_bug.md`（5 个部分），后续可单独 PR 修复。

| # | Bug | 根因 | 影响范围 | 当前处理 | pto 是否受影响 |
|---|-----|------|---------|---------|:---:|
| 1 | 3D buffer 精度异常 | codegen reduce 模板是 2D 的，3D 第 3 维被忽略 | 3 op × 2 dtype × 2 target × 2 dim = 24 case | 删除测试，文档标注"3D 暂不支持" | 是（编译失败） |
| 2 | 2D keepdim (M,1) 行归约精度错误 | `copy_ub_to_gm` codegen 对 `(M,1)` 2D buffer 模板参数行列颠倒（`<1,M>` 应为 `<M,1>`） | 3 op × 2 dtype × 2 target = 12 case | 删除测试，PR 描述反映 | 是 |
| 3 | pto BufferRegion slice 编译失败 | `GetSliceInfo` 用物理列宽算切片行数（`extent/valid_col` 误算为 2 而非 16） | 3 op × 2 dtype = 6 case | 删除测试，文档标注 pto 不支持 | N/A |
| 4 | ascendc float16 sum 列归约精度错误 | `reduce_sum_half` 的 `WholeReduceSum` `srcRepStride` 单位是 32B，无法表达 float16 列归约的 2B 步进 | 1 op × 1 dtype × 1 target = 2 case | skip，文档标注 ascendc 不支持 | 否（pto 用 `TCOLSUM` 正常） |
| 5 | ascendc float16 sum clear=False 编译失败 | `clear=False` 路径未延伸 `reduce_sum_half` workaround，直接调 `AscendC::ReduceSum<half>` 被 CANN `static_assert` 拒绝 | 1 op × 1 dtype × 1 target = 1 case | skip，文档标注 ascendc 不支持 | 否（pto 用 `TROWSUM` 正常） |

### 修复尝试说明

- Bug 2：通过 diff codegen 输出定位根因在 `T.copy` 而非 reduce，修复方向明确（修正 `copy_ub_to_gm` 模板参数行列顺序），但涉及 C++ 模板修改，未在本 PR 修复
- Bug 3：尝试在 `ReduceOpCodegen` 中修正 `slice_row`，但加 DEBUG 输出后发现该函数未被调用，pto codegen 生成 reduce 代码的真正路径未定位
- Bug 4：尝试用 cast half→float → `ReduceSum<float, RA>` → cast back 方式绕过，同样发现 `ReduceOpCodegen` 未被调用，codegen 路径未定位

### 后续待办

| 优先级 | Bug | 待办 |
|--------|-----|------|
| 高 | Bug 2 | 修 `codegen_ascend.cc` 的 `copy_ub_to_gm` 模板参数行列顺序 |
| 高 | Bug 4 | 找到 ascendc codegen 真正路径并修复 `srcRepStride` 计算 |
| 中 | Bug 5 | 延伸 `reduce_sum_half` workaround 到 `clear=False` 路径 |
| 中 | Bug 3 | 找到 pto codegen 真正路径并修复 `GetSliceInfo` 切片行数计算 |

## 文件改动

| 文件 | 类型 | 改动 |
|------|------|------|
| `tilelang/language/reduce_ascend.py` | 修改 | docstring 补充 |
| `testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py` | 新增 | 32 个测试用例 |
| `docs/api_docs/T.reduce_max.md` | 新增 | API 文档 |
| `docs/api_docs/T.reduce_min.md` | 新增 | API 文档 |
| `docs/api_docs/T.reduce_sum.md` | 新增 | API 文档 |

### CODE REVIEW

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+def test_reduce_unsupported_dtype(op, dtype):
+    """Unsupported dtype should fail to compile"""
+    N = 128
+    program = _make_1d_kernel(op, dtype, N)
+    with pytest.raises(Exception):  # noqa: B017
+        tilelang.compile(program, out_idx=[-1], pass_configs=pass_configs, target="ascendc")
+
+
+@pytest.mark.low_priority
+@pytest.mark.xfail(reason="3D dim=2 not raising; _legalize_reduce_dim should reject but doesn't")
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 分析xfail原因,  是框架问题还是说明文档要更新

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+
+@pytest.mark.low_priority
+@pytest.mark.skipif(
+    not (hasattr(torch, "npu") and torch.npu.is_available()),
+    reason="reduce correctness requires an Ascend NPU runtime",
+)
+@pytest.mark.parametrize("op", ["max", "min", "sum"])
+@pytest.mark.parametrize("dtype", ["float16", "float32"])
+@pytest.mark.parametrize("target", ["ascendc"])
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 为什么这里只有ascendc

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+
+@pytest.mark.low_priority
+@pytest.mark.skipif(
+    not (hasattr(torch, "npu") and torch.npu.is_available()),
+    reason="reduce correctness requires an Ascend NPU runtime",
+)
+@pytest.mark.parametrize("op", ["max", "min", "sum"])
+@pytest.mark.parametrize("dtype", ["float32"])
+@pytest.mark.parametrize("target", ["ascendc"])
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 为什么 target只有ascendc

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+@pytest.mark.low_priority
+@pytest.mark.parametrize("op", ["max", "min", "sum"])
+@pytest.mark.parametrize("dtype", ["int32", "bfloat16", "int16"])
+def test_reduce_unsupported_dtype(op, dtype):
+    """Unsupported dtype should fail to compile"""
+    N = 128
+    program = _make_1d_kernel(op, dtype, N)
+    with pytest.raises(Exception):  # noqa: B017
+        tilelang.compile(program, out_idx=[-1], pass_configs=pass_configs, target="ascendc")
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 为什么 target 是只 ascendc

File: docs/api_docs/T.reduce_max.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+#### 2.3.3 dim 说明
+
+dim 的合法取值取决于 buffer 的维度（rank）：
+
+| buffer rank | 合法 dim 值 | 说明 |
+|:-----------:|:----------:|------|
+| 1D | `0` / `-1` | 等价，对全部元素求最大值 |
+| 2D | `0` / `1` / `-1` / `-2` | `0`/`-2`：沿行方向归约；`1`/`-1`：沿列方向归约 |
+| 3D | `0` / `1` / `-1` / `-2` | 仅支持尾部 tile 轴（dim=0 等价于 dim=-2，dim=1 等价于 dim=-1） |
```
Comment (bingmeiyou) [inline docs/api_docs/T.reduce_max.md]: 3D这里的仅支持尾部tile轴是什么意思

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py (hunk 4 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 已修改，所有测试用例现在都覆盖了 ascendc 和 pto 两个后端（之前缺少pto）。新增发现 pto 后端对 int32/int16 缺少 dtype 校验（编译通过），已标记为 xfail，详见 PR description 的 XFAIL 详情。

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py (hunk 3 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 已修改，所有测试用例现在都覆盖了 ascendc 和 pto 两个后端（之前缺少pto）。新增发现 pto 后端对 int32/int16 缺少 dtype 校验（编译通过），已标记为 xfail，详见 PR description 的 XFAIL 详情。

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py (hunk 2 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 已修改，所有测试用例现在都覆盖了 ascendc 和 pto 两个后端（之前缺少pto）。新增发现 pto 后端对 int32/int16 缺少 dtype 校验（编译通过），已标记为 xfail，详见 PR description 的 XFAIL 详情。

File: docs/api_docs/T.reduce_max.md (hunk 5 above)
Comment (2Sheldonlaw) [reply docs/api_docs/T.reduce_max.md]: 已修改，3D dim 说明从"仅支持尾部 tile 轴"改为"仅支持最后两个维度归约，第一个维度不可归约（dim=0 归约第 2 维，dim=1 归约第 3 维）"，三个文档已同步修改。

**Review by LLMZhangYC — DISMISSED** [review]
lgtm

**Comment by cookiebai** [comment]
lgtm

**Review by fengz72 — DISMISSED** [review]

**Comment by 2Sheldonlaw** [comment]
已更新测试用例LP标签，降低ci压力

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+1. clear=False 2D 累加模式
+2. 异常边界：不支持 dtype / shape 不匹配 / dim 越界 / real_shape 错误 / rank 过高
+"""
+
+import pytest
+import tilelang
+import tilelang.language as T
+import torch
+
+tilelang.disable_cache()
```
Comment (ChaoyangJi) [inline testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 不要在模块导入时关闭全局缓存

disable_cache 修改的是进程全局 CacheState，且这里没有恢复。xdist worker 收集该模块时会执行此行，后续 fork 出来的整个测试集都会继承 cache=False，造成不必要的重复编译和超时风险。请移除此调用，或使用保存并恢复原状态的 fixture。

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    """clear=False on 2D buffer."""
+    reduce_fn = REDUCE_FNS[op]
+
+    @T.prim_func
+    def main(A: T.Tensor((M, N), dtype), B: T.Tensor((M,), dtype)):
+        with T.Kernel(1, is_npu=True) as (cid, _):
+            a_ub = T.alloc_ub((M, N), dtype)
+            b_ub = T.alloc_ub((M,), dtype)
+            T.copy(A[0, 0], a_ub)
+            T.tile.fill(b_ub, 0.0)
```
Comment (ChaoyangJi) [inline testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: clear=False 测试没有辨别力

这里把 b_ub 初始化为 0。对 sum，clear=False 与 clear=True 的结果完全相同；对每行 128 个正态随机数，max/min 也几乎必然不会被 0 改变。因此即使实现完全忽略 clear=False，该测试仍会通过，而且 PR 门禁只运行 sum/float32。请使用确定性输入及非单位初值验证 merge，例如 sum=1.25，并为 max/min 选择必然影响结果的初值。

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+# ---------------------------------------------------------------------------
+# Exception boundaries: compile-time ValueError / TypeError
+# ---------------------------------------------------------------------------
+
+
+@pytest.mark.parametrize("target", ["ascendc", pytest.param("pto", marks=pytest.mark.low_priority)])
+def test_reduce_out_shape_mismatch(target):
+    """out shape not matching reduced/keepdim shape should raise ValueError."""
+
+    with pytest.raises(Exception):  # noqa: B017
```
Comment (ChaoyangJi) [inline testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 异常测试会被无关失败误判为通过

该测试声明应产生 ValueError，但捕获了任意 Exception；DSL 解析、后端或环境的其他错误同样会让测试通过，无法证明 out shape validator 生效。后续 dim/real_shape/rank 测试也存在相同问题。请断言具体异常类型和消息，或直接对前端 validator/API 编写单元测试。

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py (hunk 6 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 已修复。

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py (hunk 7 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 已修复。输入从 torch.randn 改为 torch.arange * 0.1（确定性），初值按 op 区分：sum=1.25（期望 sum(a)+1.25，忽略 clear=False 得 sum(a)，差 1.25）、max=1e6（期望 1e6，忽略得 max(a)≈12.7）、min=-1e6（期望 -1e6，忽略得 min(a)≈-12.7）。如果 clear=False 被忽略，三种 op 都会产生明显偏差，测试有辨别力。

File: testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py (hunk 8 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_reduce_dtype_coverage.py]: 已修复。通过真机验证确认实际异常类型：bfloat16 编译失败抛 RuntimeError（match "Compilation Failed"），其余 7 个场景在 @T.prim_func 装饰阶段抛 DiagnosticError（match "DiagnosticError"）。所有 pytest.raises(Exception) 已替换为具体异常类型 + match 字符串，docstring 也同步更新为实际异常类型。

**Review by ChaoyangJi — APPROVED** [review]
lgtm

**Review by LLMZhangYC — APPROVED** [review]
approve
